doc/rados/operations/crush-map.rst

```diff
@@ -726,6 +733,21 @@ The relevant erasure-code profile properties are as follows:
    argument is omitted, then Ceph will create the CRUSH rule automatically.
 
 
+CRUSH MSR Rules
+---------------
+
+Creating an erasure-code profile with a crush-osds-per-failure-domain
+value greater than one will cause a CRUSH MSR rule type to be created
+instead of a normal CRUSH rule.  Normal crush rules cannot retry prior
+steps when an out OSD is encountered and rely on CHOOSELEAF steps to
+permit moving OSDs to new hosts.  However, CHOOSELEAF rules don't
+support more than a single OSD per failure domain.  MSR rules, new in
+squid, support multiple OSDs per failure domain by retrying all prior
+steps when an out OSD is encountered.  Using MSR rules requires that
```

Comment:
a dumb question here - if the crush rule steps are changed on-the-fly(is it possible?), and the (multi)steps gets retried, and since the initial ask was different than current values, does this impact the OSDs allocation in anyway? do we have any consequences here?

Reply:
It's possible by decompiling, editing, recompiling, and injecting the CRUSH map, like we used to have to do.

I'm far from a CRUSH expert, but I would think that so long as the old CRUSH rule and the new both fully place all PGs, the worst case would be that most or all mappings would change, and thus you'd have a thundering herd of backfill.

Reply:
See my comment above.  It seems like you are asking whether the crush rule can change during execution and the answer is no.

If you inject a new crushmap with a modification to the msr rule, it would be like changing any other crush rule -- you'd get different answers in future epochs depending on exactly what you changed.  This is not usually something we recommend people do.

@anthonyeleven's summary of the worst case consequences a above is about right.

Reply:
A few months ago I had to change the CRUSH rule for a pool that was in service but not very full.  This resulted in a bunch of OSDs OOMing.  It seems that Rook, K8s, and Ceph have unanticipated interactions wrt `osd_memory_target`, which may have been a contributing factor along with the mclock shortcomings in the Quincy release running at the time.